src/progress.rs

```diff
@@ -47,118 +47,187 @@ impl Default for ProgressState {
+        self.progress.iter_mut()
     }
 
     /// Adds a voter node
-    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        if self.voters.contains_key(&id) {
-            Err(Error::Exists(id, "voters"))?
-        }
-        if self.learners.contains_key(&id) {
-            Err(Error::Exists(id, "learners"))?;
+    pub fn insert_voter(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
+            return Err(Error::Exists(id, "voters"));
         }
-        self.voters.insert(id, pr);
+        pr.is_learner = false;
+        self.configuration.voters.insert(id);
+        self.progress.insert(id, pr);
+        debug_assert!(
+            self.configuration
+                .voters
+                .union(&self.configuration.learners)
+                .all(|v| self.progress.contains_key(v))
+        );
+        assert_eq!(
+            self.configuration.voters.len() + self.configuration.learners.len(),
+            self.progress.len()
+        );
         Ok(())
     }
 
     /// Adds a learner to the cluster
-    pub fn insert_learner(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
-        if self.voters.contains_key(&id) {
-            Err(Error::Exists(id, "voters"))?
-        }
-        if self.learners.contains_key(&id) {
-            Err(Error::Exists(id, "learners"))?
+    pub fn insert_learner(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
+        if self.learner_ids().contains(&id) {
+            return Err(Error::Exists(id, "learners"));
+        } else if self.voter_ids().contains(&id) {
+            return Err(Error::Exists(id, "voters"));
         }
-        self.learners.insert(id, pr);
+        pr.is_learner = true;
+        self.configuration.learners.insert(id);
+        self.progress.insert(id, pr);
+        debug_assert!(
+            self.configuration
+                .voters
+                .union(&self.configuration.learners)
+                .all(|v| self.progress.contains_key(v))
+        );
+        assert_eq!(
+            self.configuration.voters.len() + self.configuration.learners.len(),
+            self.progress.len()
+        );
         Ok(())
     }
 
     /// Removes the peer from the set of voters or learners.
     pub fn remove(&mut self, id: u64) -> Option<Progress> {
-        match self.voters.remove(&id) {
-            None => self.learners.remove(&id),
-            some => some,
-        }
+        self.configuration.voters.remove(&id);
+        self.configuration.learners.remove(&id);
+        let removed = self.progress.remove(&id);
+        debug_assert!(
+            self.configuration
+                .voters
+                .union(&self.configuration.learners)
+                .all(|v| self.progress.contains_key(v))
+        );
+        assert_eq!(
```

Comment:
I'm afraid the assert is not enough. What if there are `1, 2, 3` in voters and `3, 4, 5` in learners, `1, 2, 3, 4, 5, 6` in progress?

Reply:
You're right. I'll make it check both ways.